src/swap/switchain.js

```diff
@@ -0,0 +1,365 @@
+        if (lt(nativeAmount, nativeMin)) {
+          throw new SwapBelowLimitError(swapInfo, nativeMin)
+        }
+        if (gt(nativeAmount, nativeMax)) {
+          throw new SwapAboveLimitError(swapInfo, nativeMax)
+        }
+      } else {
+        const toAmountInFrom = div(toAmount, quote, 8)
+
+        if (lt(toAmountInFrom, minLimit)) {
+          throw new SwapBelowLimitError(swapInfo, nativeMin)
+        }
+
+        if (gt(toAmountInFrom, maxLimit)) {
+          throw new SwapAboveLimitError(swapInfo, nativeMax)
+        }
+      }
+
+      // order creation body
+      const quoteAmount = quoteForFrom ? { fromAmount } : { toAmount }
+      orderCreationBody = {
+        pair,
+        toAddress,
+        refundAddress: fromAddress,
+        signature,
+        orderId,
+        ...quoteAmount
+      }
+
+      // plugin output creation
+      let fromNativeAmount = nativeAmount
+      let toNativeAmount = await toWallet.denominationToNative(
+        sub(mul(quote, fromAmount), minerFee),
+        toCurrencyCode
+      )
+      if (!quoteForFrom) {
+        fromNativeAmount = await fromWallet.denominationToNative(
+          div(add(toAmount, minerFee), quote, 8),
+          fromCurrencyCode
+        )
+        toNativeAmount = nativeAmount
+      }
+
+      const destinationAddress = toAddress
+      const expirationDate = new Date(expiryTs * 1000)
+
+      // create preliminary tx using our recieve address to calculate a networkFee
+      const spendInfo: EdgeSpendInfo = {
+        currencyCode: fromCurrencyCode,
+        spendTargets: [
+          { nativeAmount: fromNativeAmount, publicAddress: fromAddress }
+        ]
+      }
+
+      const preliminaryTx: EdgeTransaction = await fromWallet.makeSpend(
+        spendInfo
+      )
+
+      // convert that to the output format:
+      return makeSwapPluginQuote(
+        request,
+        fromNativeAmount,
+        toNativeAmount,
+        preliminaryTx,
+        destinationAddress,
+        swapInfo.pluginName,
+        false,
+        expirationDate,
+        orderId
+      )
+    }
+  }
+
+  function makeSwapPluginQuote(
+    request: EdgeSwapRequest,
+    fromNativeAmount: string,
+    toNativeAmount: string,
+    tx: EdgeTransaction,
+    destinationAddress: string,
+    pluginId: string,
+    isEstimate: boolean = false,
+    expirationDate?: Date,
+    quoteId?: string
+  ): EdgeSwapPluginQuote {
+    const { fromWallet } = request
+
+    const out: EdgeSwapPluginQuote = {
+      fromNativeAmount,
+      toNativeAmount,
+      networkFee: {
+        currencyCode: fromWallet.currencyInfo.currencyCode,
+        nativeAmount: tx.networkFee
+      },
+      destinationAddress,
+      pluginName: swapInfo.pluginName,
+      expirationDate,
+      quoteId,
+      isEstimate,
+      async approve(): Promise<EdgeTransaction> {
+        const json: SwitchainOrderCreationResponse = await swHttpCall(
+          '/order',
+          'POST',
+          orderCreationBody
```

Comment:
Actually, this looks like a potential race condition, since we don't know that the `orderCreationBody` set up in the quote is necessarily the same one we want to approve (maybe the GUI gets a bug one day where it calls `quote` a bunch of times, leaving the wrong value in this variable and causing users to lose money). Could you pass this into `makeSwapPluginQuote` as a parameter, so we guarantee a 1:1 connection between quoting and approval?

Reply:
Moved `orderCreationBody` to `makeSwapPluginQuote` params as per your request.